<?xml version="1.0"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro">

    <xacro:include filename="$(find sentry_kinematics_sim)/xacro/sentry_laser.xacro" />

  <xacro:property name="M_PI" value="3.14159"/> 
  <xacro:property name="laser_min_range" value="0.1"/>
  <xacro:property name="laser_max_range" value="200.0"/>
  <xacro:property name="ros_topic" value="scan"/>
  <xacro:property name="samples" value="10000"/>
  <xacro:property name="downsample" value="0"/>
  
  <xacro:macro name="null_inertial">
    <inertial>
      <mass value="0.1"/>
    <inertia ixx="0.01" ixy="0" ixz="0"
          iyy="0.01" iyz="0"
          izz="0.01"/>
    </inertial>

  </xacro:macro>
  <xacro:macro name="Livox_Mid_gazebo_sensor" params="visualize:=True update_rate:=10 resolution:=0.002 noise_mean:=0.0 noise_stddev:=0.01 name:=livox">
    <gazebo reference="${name}_Link">
      <sensor type="ray" name="livox_${name}">
        <pose>0 0 0 0 0 0</pose>
        <visualize>${visualize}</visualize>
        <update_rate>10</update_rate>
        <plugin name="gazebo_ros_laser_controller" filename="liblivox_laser_simulation.so">
			<ray>
			  <scan>
				<horizontal>
				<samples>100</samples>
				<resolution>1</resolution>
				<min_angle>${0}</min_angle>
				<max_angle>${2*M_PI}</max_angle>
				</horizontal>
				<vertical>
				<samples>50</samples>
				<resolution>1</resolution>
				<min_angle>${-7.22/180*M_PI}</min_angle>
				<max_angle>${55.22/180*M_PI}</max_angle>
				</vertical>
			  </scan>
			  <range>
				<min>${laser_min_range}</min>
				<max>${laser_max_range}</max>
				<resolution>${resolution}</resolution>
			  </range>
			  <noise>
				<type>gaussian</type>
				<mean>${noise_mean}</mean>
				<stddev>${noise_stddev}</stddev>
			  </noise>
			</ray>
          <visualize>${visualize}</visualize>
		  <samples>${samples}</samples>
		  <downsample>${downsample}</downsample>
		  <csv_file_name>package://sentry_laser_sim/scan_mode/mid360.csv</csv_file_name>
		  <ros_topic>${ros_topic}</ros_topic>
        </plugin>
      </sensor>
    </gazebo>
  </xacro:macro>

<!--
    <link name="livox">
      <xacro:null_inertial/>
    </link>

    <joint name="livox_to_Laser_Link_joint" type="fixed">
      <parent link="Laser_Link"/>
      <child link="livox"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
    </joint>
-->


    <xacro:Livox_Mid_gazebo_sensor name="Laser" visualize="true"/>

</robot>
